File: src/main/java/kr/co/bookand/backend/issue/service/IssueService.java
package kr.co.bookand.backend.issue.service;

import kr.co.bookand.backend.account.domain.Account;
import kr.co.bookand.backend.account.service.AccountService;
import kr.co.bookand.backend.common.domain.Message;
import kr.co.bookand.backend.common.domain.dto.PageResponse;
import kr.co.bookand.backend.issue.domain.Issue;
import kr.co.bookand.backend.issue.domain.IssueImage;
import kr.co.bookand.backend.issue.repository.IssueImageRepository;
import kr.co.bookand.backend.issue.repository.IssueRepository;
import kr.co.bookand.backend.util.s3.service.AwsS3Service;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

import static kr.co.bookand.backend.issue.domain.dto.IssueDto.*;
import static kr.co.bookand.backend.util.s3.dto.AwsS3Dto.*;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
@Slf4j
public class IssueService {

    private final IssueRepository issueRepository;
    private final IssueImageRepository issueImageRepository;
    private final AccountService accountService;
    private final AwsS3Service awsS3Service;


    @Transactional
    public IssueIdResponse createIssue (CreateIssueRequest createIssueRequest) {
        Account loginAccount = accountService.getCurrentAccount();
        accountService.isAccountAdmin();
        List<String> issueImages = createIssueRequest.issueImages();


        FileDto fileDto = null;
        if (createIssueRequest.logFile() != null) {
            MultipartFile multipartFile = createIssueRequest.logFile();
            fileDto = awsS3Service.uploadV2(multipartFile, "reportLog", loginAccount.getId().toString());
        }
        Issue issue = createIssueRequest.toEntity(loginAccount.getId());

        if (createIssueRequest.sendLogs()) {
            issue.setLogFile(fileDto.fileUrl(), fileDto.filename());
        }

        Issue saveIssue = issueRepository.save(issue);
        List<IssueImage> issueImageList = issueImages
                .stream()
                .map(issueImage -> {
                    IssueImage image = IssueImage.builder().imageUrl(issueImage).issue(saveIssue).build();
                    IssueImage saveIssueImage = issueImageRepository.save(image);
                    return saveIssueImage;
                })
                .toList();
        issue.setIssueImage(issueImageList);

        return IssueIdResponse.of(issue);
    }

    public IssueResponse getIssue(Long issueId) {
        Issue issue = issueRepository.findById(issueId).orElseThrow();
        return IssueResponse.of(issue);
    }

    public PageResponse<IssueSimpleResponse> getIssueList(Pageable pageable) {
        accountService.isAccountAdmin();
        return PageResponse.of(issueRepository.findAll(pageable).map(IssueSimpleResponse::of));
    }

    @Transactional
    public Message checkConfirmed(Long issueId, boolean checkConfirmed) {
        Issue issue = issueRepository.findById(issueId).orElseThrow();
        issue.checkConfirmed(checkConfirmed);
        return Message.of("오류 및 버그 신고 체크 완료");
    }
}

File: src/main/kotlin/kr/co/bookand/backend/issue/service/IssueService.kt
package kr.co.bookand.backend.issue.service

import kr.co.bookand.backend.account.model.Account
import kr.co.bookand.backend.common.ErrorCode
import kr.co.bookand.backend.common.PageResponse
import kr.co.bookand.backend.common.exception.BookandException
import kr.co.bookand.backend.issue.dto.*
import kr.co.bookand.backend.issue.model.Issue
import kr.co.bookand.backend.issue.model.IssueImage
import kr.co.bookand.backend.issue.repository.IssueImageRepository
import kr.co.bookand.backend.issue.repository.IssueRepository
import kr.co.bookand.backend.util.s3.dto.FileDto
import kr.co.bookand.backend.util.s3.service.AwsS3Service
import lombok.RequiredArgsConstructor
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
class IssueService(
    private val issueRepository: IssueRepository,
    private val issueImageRepository: IssueImageRepository,
    private val awsS3Service: AwsS3Service
) {

    fun getIssue(issueId: Long): IssueResponse {
        return issueRepository.findById(issueId)
            .map { IssueResponse(it) }
            .orElseThrow { throw BookandException(ErrorCode.NOT_FOUND_ISSUE) }
    }

    fun getIssueList(
        currentAccount: Account,
        pageable: Pageable
    ): IssueSimpleListResponse {
        currentAccount.role.checkAdminAndManager()
        val map = issueRepository.findAll(pageable)
            .map { IssueSimpleResponse(it) }
        return IssueSimpleListResponse(PageResponse.of(map))
    }

    @Transactional
    fun createIssue(
        currentAccount: Account,
        createIssueRequest: CreateIssueRequest
    ): IssueIdResponse {
        val issueImagesList = createIssueRequest.issueImages

        var fileDto = FileDto("", "")
        if (createIssueRequest.logFile != null) {
            val multipartFile = createIssueRequest.logFile
            fileDto = awsS3Service.uploadV2(multipartFile, "reportLog", owner = currentAccount.email)
        }
        val issue = Issue(createIssueRequest, 1L)

        if (createIssueRequest.sendLogs) {
            issue.setLogFile(fileDto.fileUrl, fileDto.filename)
        }

        val saveIssue: Issue = issueRepository.save(issue)
        val issueImageList = issueImagesList
            .map { issueImage: String ->
                val image = IssueImage(issueImage, saveIssue)
                val saveIssueImage: IssueImage = issueImageRepository.save(image)
                saveIssueImage
            }
            .toList()
        issue.setIssueImage(issueImageList)

        return IssueIdResponse(saveIssue.id)
    }

    @Transactional
    fun updateConfirmed(
        currentAccount: Account,
        issueId: Long,
        updateConfirmed: Boolean
    ): IssueIdResponse {
        currentAccount.role.checkAdminAndManager()
        val issue: Issue = issueRepository.findById(issueId)
            .orElseThrow { throw BookandException(ErrorCode.NOT_FOUND_ISSUE) }
        issue.updateConfirmed(updateConfirmed)
        return IssueIdResponse(issue.id)
    }
}